libs/providers/config-cat/src/lib/config-cat-provider.ts

```diff
@@ -1,84 +1,80 @@
   public readonly events = new OpenFeatureEventEmitter();
-  private _status: ProviderStatus = ProviderStatus.NOT_READY;
-  private readonly _clientParameters: Parameters<typeof getClient>;
+  private readonly _clientFactory: (provider: ConfigCatProvider) => IConfigCatClient;
+  private _hasError = false;
   private _client?: IConfigCatClient;
 
+  public runsOn: Paradigm = 'server';
+
   public metadata = {
     name: ConfigCatProvider.name,
   };
 
-  constructor(...params: Parameters<typeof getClient>) {
-    this._clientParameters = params;
+  protected constructor(clientFactory: (provider: ConfigCatProvider) => IConfigCatClient) {
+    this._clientFactory = clientFactory;
   }
 
-  public static create(...params: Parameters<typeof getClient>) {
-    return new ConfigCatProvider(...params);
-  }
-
-  get status() {
-    return this._status;
-  }
-
-  public async initialize(): Promise<void> {
-    return new Promise((resolve) => {
-      const originalParameters = this._clientParameters;
-      originalParameters[2] ??= {};
-
-      const options = originalParameters[2];
-      const oldSetupHooks = options.setupHooks;
+  public static create<TMode extends PollingMode>(
+    sdkKey: string,
+    pollingMode?: TMode,
+    options?: OptionsForPollingMode<TMode>,
+  ): ConfigCatProvider {
+    // Let's create a shallow copy to not mess up caller's options object.
+    options = options ? { ...options } : ({} as OptionsForPollingMode<TMode>);
+    return new ConfigCatProvider((provider) => {
+      const oldSetupHooks = options?.setupHooks;
 
       options.setupHooks = (hooks) => {
         oldSetupHooks?.(hooks);
 
-        // After resolving, once, we can simply emit events the next time
-        const onProviderReady = () => {
-          this.events.emit(ProviderEvents.Ready);
-          this._status = ProviderStatus.READY;
-        };
-
-        hooks.once('clientReady', () => {
-          hooks.on('clientReady', onProviderReady);
-          onProviderReady();
-          resolve();
-        });
-
         hooks.on('configChanged', (projectConfig: IConfig | undefined) =>
-          this.events.emit(ProviderEvents.ConfigurationChanged, {
+          provider.events.emit(ProviderEvents.ConfigurationChanged, {
             flagsChanged: projectConfig ? Object.keys(projectConfig.settings) : undefined,
           }),
         );
 
         hooks.on('clientError', (message: string, error) => {
-          this._status = ProviderStatus.ERROR;
-          this.events.emit(ProviderEvents.Error, {
+          provider._hasError = true;
+          provider.events.emit(ProviderEvents.Error, {
             message: message,
             metadata: error,
           });
         });
       };
 
-      this._client = getClient(...originalParameters);
+      return getClient(sdkKey, pollingMode, options);
     });
   }
 
+  public async initialize(): Promise<void> {
+    const client = this._clientFactory(this);
+    await client.waitForReady();
```

Comment:
This makes init easy for this provider... just a point in our discussion about "auto-ready-events".